DOC: add key concepts doc

jakevdp opened the pull request:
This will replace the new content in thinking-in-jax within the new tutorial flow.

Followup to #20306

Rendered version will appear at: https://jax--20347.org.readthedocs.build/en/20347/tutorials/key-concepts.html

skye commented on docs/glossary.rst:
```diff
@@ -3,6 +3,9 @@ JAX Glossary of Terms
 .. glossary::
 
+    Array
+      JAX's analog of the :class:`numpy.ndarray`. See :class:`jax.Array`.
```
```suggestion
      JAX's analog of :class:`numpy.ndarray`. See :class:`jax.Array`.
```

jakevdp replied:
Done

skye commented on docs/tutorials/key-concepts.md:
```diff
@@ -0,0 +1,192 @@
+(key-concepts)=
+# Key Concepts
+
+This sectiong briefly introduces some key concepts of the JAX package.
```
```suggestion
This section briefly introduces some key concepts of the JAX package.
```

jakevdp replied:
Done

skye commented on docs/tutorials/key-concepts.md:
```diff
@@ -0,0 +1,192 @@
+## Tracing
+
+The magic behind transformations is the notion of a {term}`Tracers <Tracer>`.
+Tracers are abstract standins for array objects, and are passed to JAX functions in order
```
```suggestion
Tracers are abstract stand-ins for array objects, and are passed to JAX functions in order
```

jakevdp replied:
Done

jakevdp committed:
DOC: add key concepts doc

This will replace the new content in thinking-in-jax within the new tutorial flow.
